### DISCUSSION TITLE
about motion_gen

### DISCUSSION THREAD

**Discussion by Spphire** [NVlabs/curobo#25] [Software Q&A]
Is it possible to set the velocity of init point and end point of planning? for better connect the 2 plans before and after

**Comment by balakumar-s** [comment]
That's currently not possible in the motion generation pipeline as it assumes zero velocity at start and end points to perform the optimization. 

You can create your own trajectory optimization problem to perform optimization with non-zero start and end velocities.

**Comment by Spphire** [comment]
Is there any exemple or API can be used to create?

**Reply by balakumar-s** [reply]
There is  no example on how to do this. This would be the file to look at: https://github.com/NVlabs/curobo/blob/main/src/curobo/wrap/reacher/trajopt.py

**Reply by Spphire** [reply]
https://github.com/NVlabs/curobo/blob/e0e804e9062da33c13aa262a1da253d364b8768a/src/curobo/rollout/arm_reacher.py#L207
maybe add some velocity-cost in "cost_fn" function

**Reply by Spphire** [reply]
https://github.com/NVlabs/curobo/blob/e0e804e9062da33c13aa262a1da253d364b8768a/src/curobo/rollout/arm_reacher.py#L302
found it

**Reply by balakumar-s** [reply]
non-zero velocity at start state might work with the existing api. However, non-zero velocity at final timestep will require rewriting the rollout class. Did you try setting non-zero velocity for the start state?

**Reply by Spphire** [reply]
if just for the start state, which API？

**Reply by balakumar-s** [reply]
We set start velocity to zero here: https://github.com/NVlabs/curobo/blob/e0e804e9062da33c13aa262a1da253d364b8768a/examples/isaac_sim/motion_gen_reacher.py#L239

You can instead send a non-zero velocity for the start state.

**Reply by Spphire** [reply]
ill try it

**Comment by Spphire** [comment]
> We set start velocity to zero here:
> 
> https://github.com/NVlabs/curobo/blob/e0e804e9062da33c13aa262a1da253d364b8768a/examples/isaac_sim/motion_gen_reacher.py#L239
> 
> You can instead send a non-zero velocity for the start state.
emm, i edited it like "velocity=tensor_args.to_device(sim_js.velocities),", but it occured frequently warning of "plan didnt converge to a solution"

**Reply by balakumar-s** [reply]
Optimizing with non-zero start velocity is a harder optimization problem so you will see more failures. Also the pipeline is mainly setup for zero velocity at start and end. 

What kind of application are you trying to use this pipeline for that has non-zero start and end states?

**Comment by Spphire** [comment]
I found that in motion_gen_reacher.py, there are still some slight pauses when the robot tracking the cube. (slow down, then re-follow the new target)
I want to make a more smooth version

**Reply by balakumar-s** [reply]
You will see pauses as it's obtaining a full motion plan and executing. If you want a more immediate following example, see this https://github.com/NVlabs/curobo/blob/main/examples/isaac_sim/mpc_example.py . This uses model predictive control to track the cube in a reactive setting. However, mpc is local and it can get stuck in local minima.

**Reply by Spphire** [reply]
nice job! ill try it
